Enabled third party auth from account settings page.

symbolist opened the pull request:
Sandbox: http://userprofile.m.sandbox.edx.org/account/settings

The third section only shows if third party auth is enabled.
And the list is dynamic i.e. it will allow linking/unlinking with all the configured list of authentication providers.

symbolist commented:
@explorerleslie @srpearce @marcotuts This PR adds support for third party auth to the account settings page. I have update the existing sandbox to include this functionality too.

http://userprofile.m.sandbox.edx.org/account/settings

symbolist commented:
@stephensanchez Can you give a quick look to see that the mechanism is right?

stephensanchez commented:
The code looks great. One overall comment, is to add a lot of comments to the pipeline changes and defining your javascript methods. The 3rd party auth pipeline can be pretty confusing, and explaining how you get returned to the accounts profile page would be valuable in my opinion. 

So if you add doc to all your new python and javascript, I'd say ship it. :+1:

explorerleslie commented:
@symbolist Google Linking looks good. When I try to link Facebook, I get the following message -- any idea what's going on?

![image](https://cloud.githubusercontent.com/assets/5643426/6876262/1d4aeda8-d49c-11e4-9557-72f6804bff7d.png)

symbolist commented:
@stephensanchez I have added some docs to the new Python and Javascript code.

@explorerleslie I am probably not configuring the settings in the Facebook developer account right. I tried fiddling with things but not sure if that resolves the issue for you.

But in any case, as long as a provider is configured correctly in the edX instance settings, it should behave exactly like how the Google auth is working and how linking works on the dashboard.

@muzaffaryousaf Can you also review this? I have added the tests too.

symbolist commented:
@explorerleslie It looks like it was because I hadn't changed the configuration in the Facebook account from developer to live mode. I have done that now so hopefully it should work for you now too.

muzaffaryousaf commented:
@symbolist There was a duplicate provider check in previous code https://github.com/edx/edx-platform/blob/master/common/djangoapps/student/views.py#L697 that shows the error message in case of account being already used. I think we would be needing that here as well.

muzaffaryousaf commented on lms/static/js/student_account/views/account_settings_fields.js:
```diff
@@ -99,6 +99,74 @@
+
+        });
+
+        AccountSettingsFieldViews.AuthFieldView = FieldViews.LinkFieldView.extend({
```
Implementation of  this feature by using Fields is really awesome.

muzaffaryousaf commented:
Once you address duplicate provider thing: :+1:

symbolist commented:
test this please.

symbolist commented:
@muzaffaryousaf Thanks for the catch! I have added that in too.

@explorerleslie @marcotuts If the user tries to connect to an account that is already connected with another edX account, they get a message like this:

![screen shot 2015-03-29 at 17 57 44](https://cloud.githubusercontent.com/assets/5305552/6887251/0fe68bda-d673-11e4-9f31-d251a582d419.png)

This is exactly how it looks on the dashboard. To produce this on the sandbox:
1. Sign in with honor user and link one of Facebook or Google.
2. Sign in with staff user and try to link again with the same provider.

muzaffaryousaf commented:
@symbolist can we add/update python unit tests for this feature just like they were added before. You can see them in 2 files. 
1- https://github.com/edx/edx-platform/blob/master/common/djangoapps/third_party_auth/tests/specs/test_google.py

2- https://github.com/edx/edx-platform/blob/master/common/djangoapps/third_party_auth/tests/specs/test_linkedin.py

symbolist commented:
test this please.

symbolist commented:
@muzaffaryousaf I was assuming you would redirect the tests when you made changes to the dashboard. But I can make them here.

You can rebase that branch when I am done.

symbolist commented:
test this please.

explorerleslie commented:
:+1:

marcotuts commented:
(I wasn't able to get to the sandbox but I'm to review the dashboard error message later. Thanks for the image and context in the PR, it was very helpful!) :+1:

symbolist commented:
@nasthagiri This is the PR in which the third-party auth UI was added to the account settings page. Do you think I missed something?

nasthagiri commented:
@symbolist Is there a sandbox with this feature that I can play with?

nasthagiri commented on common/djangoapps/third_party_auth/tests/specs/base.py:
```diff
@@ -432,7 +458,7 @@ def test_full_pipeline_succeeds_for_linking_account(self):
 
         # First we expect that we're in the unlinked state, and that there
         # really is no association in the backend.
-        self.assert_dashboard_response_looks_correct(student_views.dashboard(request), request.user, linked=False)
```
Please delete `assert_dashboard_response_looks_correct` method since it is no longer used.

nasthagiri commented on common/djangoapps/third_party_auth/pipeline.py:
```diff
@@ -117,6 +117,7 @@ def B(*args, **kwargs):
 AUTH_ENTRY_DASHBOARD = 'dashboard'
 AUTH_ENTRY_LOGIN = 'login'
 AUTH_ENTRY_REGISTER = 'register'
+AUTH_ENTRY_ACCOUNT_SETTINGS = 'account_settings'
```
Are we no longer supporting `AUTH_ENTRY_DASHBOARD`?  If so, we should remove it.

nasthagiri commented on common/test/acceptance/tests/lms/test_account_settings.py:
```diff
@@ -254,3 +254,17 @@ def test_prefered_language_field(self):
             [u'Pushto', u''],
         )
+
+    def test_connected_accounts(self):
```
These should exist only if the feature flag for 3rd party auth is enabled, correct?

symbolist replied:
@nasthagiri Yes, the section only shows if third party auth is enabled.

symbolist commented:
@nasthagiri Sandbox: http://mzfr-profile-photo.m.sandbox.edx.org/

nasthagiri commented:
@symbolist I don't see in this PR where the old 3rd-party-auth code on the dashboard was removed.  For example, shouldn't dashboard.html be cleaned up since the UI is no longer there?

nasthagiri commented:
The wording in this dialog needs to be updated (since the user needs to be directed to the account-settings page instead of the dashboard).

![screen shot 2015-04-16 at 2 26 43 pm](https://cloud.githubusercontent.com/assets/6772826/7188326/e5a34aea-e444-11e4-8f41-b8c20f559a40.png)

symbolist commented:
@nasthagiri Here is the PR which cleaned up the dashboard: https://github.com/edx/edx-platform/pull/7506
